core/node/node_sync/src/tree_data_fetcher/provider/mod.rs

```diff
@@ -0,0 +1,321 @@
+                )
+                .await?;
+                tracing::debug!(
+                    number = number.0,
+                    "Guessed L1 block number for L1 batch #{number} commit: {approximate_block}"
+                );
+                // Subtract to account for imprecise L1 and L2 timestamps etc.
+                approximate_block.saturating_sub(Self::L1_BLOCK_ACCURACY)
+            }
+        };
+
+        let number_topic = H256::from_low_u64_be(number.0.into());
+        let filter = web3::FilterBuilder::default()
+            .address(vec![self.diamond_proxy_address])
+            .from_block(web3::BlockNumber::Number(from_block))
+            .to_block(web3::BlockNumber::Number(from_block + Self::L1_BLOCK_RANGE))
+            .topics(
+                Some(vec![self.block_commit_signature]),
+                Some(vec![number_topic]),
+                None,
+                None,
+            )
+            .build();
+        let mut logs = self.eth_client.logs(&filter).await?;
+        logs.retain(|log| !log.is_removed() && log.block_number.is_some());
+
+        match logs.as_slice() {
+            [] => Ok(Err(MissingData::Batch)),
+            [log] => {
+                let root_hash_topic = log.topics.get(2).copied().ok_or_else(|| {
+                    let err = "Bogus `BlockCommit` event, does not have the root hash topic";
+                    EnrichedClientError::new(ClientError::Custom(err.into()), "batch_details")
+                        .with_arg("filter", &filter)
+                        .with_arg("log", &log)
+                })?;
+                // `unwrap()` is safe due to the filtering above
+                let l1_commit_block_number = log.block_number.unwrap();
+
+                let l1_commit_block = self.eth_client.block(l1_commit_block_number.into()).await?;
+                let l1_commit_block = l1_commit_block.ok_or_else(|| {
+                    let err = "Block disappeared from L1 RPC provider";
+                    EnrichedClientError::new(ClientError::Custom(err.into()), "batch_details")
+                        .with_arg("number", &l1_commit_block_number)
+                })?;
+                self.past_l1_batch = Some(PastL1BatchInfo {
+                    number,
+                    l1_commit_block_number,
+                    l1_commit_block_timestamp: l1_commit_block.timestamp,
+                });
+                Ok(Ok(root_hash_topic))
+            }
+            _ => {
+                tracing::warn!("Non-unique `BlockCommit` event for L1 batch #{number} queried using {filter:?}: {logs:?}");
+                Ok(Err(MissingData::RootHash))
+            }
+        }
+    }
+}
+
+/// Data provider combining [`L1DataProvider`] with a fallback provider.
+#[derive(Debug)]
+pub(super) struct CombinedDataProvider {
+    l1: Option<L1DataProvider>,
+    fallback: Box<dyn TreeDataProvider>,
+}
+
+#[async_trait]
+impl TreeDataProvider for CombinedDataProvider {
```

Comment:
A nice property is that any errors emitted by `L1DataProvider` are not treated as fatal; they either lead to switching it off, or to using the L2 provider as a fallback (in case of transient errors). I.e., `L1DataProvider` is only required to not return bogus root hashes, which I *think* is the case.